x/evm/keeper/state_transition.go

```diff
@@ -191,6 +191,9 @@ func (k *Keeper) ApplyTransaction(tx *ethtypes.Transaction) (*types.MsgEthereumT
 		return nil, stacktrace.Propagate(err, "failed to apply ethereum core message")
 	}
 
+	// flatten the cache contexts to improve efficiency of following db operations
+	k.ctxStack.CommitToRevision(revision)
```

Comment:
how does this relate to `k.CommitCachedContexts()`? why is this called here instead of inside the `!res.Failed` condition? I need more context to understand the rationale of this change

Reply:
For more context, we discovered a bug in our testnet which make the network to halt.

Some uniswap call triggers a lot of internal call. And trying to replay the tx will end up in an infinite loop there
https://github.com/cosmos/cosmos-sdk/blob/master/store/cachekv/mergeiterator.go#L206

Reply:
The change here doesn't provide context on the following questions:

- How does flattening improve efficiency? Do you have benchmark results?
- Is it just to commit changes and avoid the infinite loop? 
- What's the threshold (i.e max number) of internal calls that the EVM supports without crashing? 
- Can we add a test to check the max number of internal calls supported?
- Do we still need to call the `CommitCachedContexts()` if the changes are being committed regardless if it fails or not? Can this logic commit an invalid result or affect the revert logic as well?

Since the underlying issue is from the SDK, you should open an issue there to get it fixed in the long term

Reply:
![Screenshot from 2021-10-03 19-34-25](https://user-images.githubusercontent.com/51560997/136060166-023fdf84-ce0a-4cbb-9246-c1b9e32e1f1b.png)

it hangs forever here,
```
// Commit commits all the cached contexts from top to bottom in order and clears the stack by setting an empty slice of cache contexts.
func (cs *ContextStack) Commit() {
	// commit in order from top to bottom
	for i := len(cs.cachedContexts) - 1; i >= 0; i-- {
		// keep all the cosmos events
		cs.initialCtx.EventManager().EmitEvents(cs.cachedContexts[i].ctx.EventManager().Events())
		if cs.cachedContexts[i].commit == nil {
			panic(fmt.Sprintf("commit function at index %d should not be nil", i))
		} else {
			cs.cachedContexts[i].commit()
		}
	}
	cs.cachedContexts = []cachedContext{}
}
```

this function calls forever, i think , in some callstack contains itself?

Reply:
So in this pr, before calling Commit,   calls cached contexts Commit